packages/dht/src/dht/PeerManager.ts

```diff
@@ -0,0 +1,339 @@
+                contact.ping().then((result) => {
+                    if (result) {
+                        logger.trace(`Added new contact ${contact.getPeerId().value.toString()}`)
+                        this.emit(
+                            'newKbucketContact',
+                            contact.getPeerDescriptor(),
+                            this.neighborList!.getClosestContacts(this.config.getClosestContactsLimit).map((peer) => peer.getPeerDescriptor())
+                        )
+                    } else {
+                        logger.trace('ping failed ' + this.config.nodeName + ', ' + contact.getPeerDescriptor().nodeName + ' ')
+                        this.config.connectionManager?.weakUnlockConnection(contact.getPeerDescriptor())
+                        this.removeContact(contact.getPeerDescriptor())
+                        this.addClosestContactToBucket()
+                    }
+                    return
+                }).catch((_e) => {
+                    this.config.connectionManager?.weakUnlockConnection(contact.getPeerDescriptor())
+                    this.removeContact(contact.getPeerDescriptor())
+                    this.addClosestContactToBucket()
+                })
+            }
+        }
+    }
+
+    private addClosestContactToBucket(): void {
+        if ( this.stopped ) {
+            return
+        }
+        const closest = this.getClosestActiveContactNotInBucket()
+        if (closest) {
+            this.handleNewPeers([ closest.getPeerDescriptor() ])
+        }
+    }
+
+    private getClosestActiveContactNotInBucket(): DhtPeer | undefined {
+        for (const contactId of this.neighborList!.getContactIds()) {
+            if (!this.bucket!.get(contactId.value) && this.neighborList!.isActive(contactId)) {
+                return this.neighborList!.getContact(contactId).contact
+            }
+        }
+        return undefined
+    }
+
+    public handleConnected(peerDescriptor: PeerDescriptor): void {
+        if (this.config.ownPeerId!.equals(PeerID.fromValue(peerDescriptor.kademliaId))) {
+            logger.error('handleConnected() to self')
+            return
+        }
+
+        const dhtPeer = this.config.createDhtPeer(peerDescriptor)
+        if (!this.connections.has(PeerID.fromValue(dhtPeer.id).toKey())) {
+            this.connections.set(PeerID.fromValue(dhtPeer.id).toKey(), dhtPeer)
+            logger.trace(' ' + this.config.nodeName + ' connectionschange add ' + this.connections.size)
+        } else {
+            logger.trace('new connection not set to connections, there is already a connection with the peer ID')
+        }
+        if (this.config.nodeName === 'entrypoint') {
+            logger.trace('connected: ' + this.config.nodeName + ', ' + peerDescriptor.nodeName + ' ' + this.connections.size)
+        }
+    }
+
+    public handleDisconnected(peerDescriptor: PeerDescriptor, disconnectionType: DisconnectionType, isLayer0 = false): void {
```

Comment:
Maybe good to have the `isLayer0` as a required parameter as it is not trivial to the caller what would be the default. Alternatively this could be moved to a field of `PeerManager` if the value never changes during the lifecycle of the instance.

```suggestion
    public handleDisconnected(peerDescriptor: PeerDescriptor, disconnectionType: DisconnectionType, isLayer0: boolean): void {
```

Reply:
moved to PeerManagerConfig